### PR TITLE
luminous: os/bluestore: backport new bitmap allocator

### PR DESCRIPTION
**Pull request by ifed01**
https://tracker.ceph.com/issues/38778

### CODE REVIEW

**Comment by xiexingguo** [comment]
@ifed01 Do we have any performance bench data showing that the new fast bitmap allocator is comparable to (or better than) the old stupid allocator?

**Comment by xiaoxichen** [comment]
@xiexingguo  checkout http://lists.ceph.com/pipermail/ceph-users-ceph.com/2019-January/032737.html for background.

in short, stupid allocator is stupid in some cases , b-tree get fragmented heavily causing significant latency spike.

**Comment by xiexingguo** [comment]
@

**Comment by ifed01** [comment]
> @ifed01 Do we have any performance bench data showing that the new fast bitmap allocator is comparable to (or better than) the old stupid allocator?

@xiexingguo - here are some slides about bitmap allocator I presented a while ago:
https://docs.google.com/presentation/d/1_1Otkgv76fbCU2Zogjz748sEAG-1Nfiidbb6mgTON-A/edit?usp=sharing

File: src/common/options.cc [hunk 1]
```diff
@@ -3655,7 +3655,7 @@ std::vector<Option> get_global_options() {
     .set_description("Key value database to use for bluestore"),
 
     Option("bluestore_allocator", Option::TYPE_STR, Option::LEVEL_DEV)
-    .set_default("stupid")
+    .set_default("bitmap")
```
Comment (liewegas) [inline src/common/options.cc]: Perhaps for the intiial backport we leave teh default as stupid and let some users opt in?

I'm a bit worried there will be some weird pathological allocation behavior that we don't see in qa but happens with production workloads.  The new allocator is very stable in qa but it hasn't actually ever been deployed in a real cluster yet.

Comment (ifed01) [reply src/common/options.cc]: yeah, makes sense. Reverting...

**Comment by yuriw** [comment]
[wip-yuri4-testing-2019-03-20-2018-luminous](https://trello.com/c/BwpOdBxX/564-wip-yuri4-testing-2019-03-20-2018-luminous)

**Comment by yuriw** [comment]
@ifed01 This PR seems breaking tests' yaml configuration:

teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ceph/ceph-ci.git -c $CEPH_BRANCH -m $MACHINE_NAME -s rados --subset 35/999 -k distro -p 90 -e $CEPH_QA_MAIL --suite-branch $CEPH_BRANCH
2019-03-21 15:24:18,578.578 INFO:teuthology.suite:Using random seed=7603
2019-03-21 15:24:18,579.579 INFO:teuthology.suite.run:kernel sha1: distro
2019-03-21 15:24:18,828.828 DEBUG:teuthology.repo_utils:git ls-remote ￼ceph/ceph-ci wip-yuri4-testing-2019-03-20-2018-luminous -> 91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:18,828.828 INFO:teuthology.suite.run:ceph sha1: 91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:18,829.829 DEBUG:teuthology.suite.util:Defaults for machine_type smithi distro centos: arch=x86_64, release=centos/7, pkg_type=rpm
2019-03-21 15:24:18,830.830 DEBUG:teuthology.packaging:Querying https://shaman.ceph.com/api/search?status=ready&project=ceph&flavor=default&distros=centos%2F7%2Fx86_64&sha1=91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:19,019.019 INFO:teuthology.suite.run:ceph version: 12.2.11-277.g9105290
2019-03-21 15:24:19,019.019 INFO:teuthology.suite.run:teuthology branch: master
2019-03-21 15:24:19,293.293 DEBUG:teuthology.repo_utils:git ls-remote https://github.com/ceph/ceph-ci.git wip-yuri4-testing-2019-03-20-2018-luminous -> 91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:19,543.543 DEBUG:teuthology.repo_utils:git ls-remote https://github.com/ceph/ceph-ci.git wip-yuri4-testing-2019-03-20-2018-luminous -> 91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:19,543.543 INFO:teuthology.suite.run:ceph-ci branch: wip-yuri4-testing-2019-03-20-2018-luminous 91052903236ec470cda8b505f10417ebfb9f3f6d
2019-03-21 15:24:19,545.545 DEBUG:teuthology.repo_utils:Setting repo remote to https://github.com/ceph/ceph-ci.git
2019-03-21 15:24:19,552.552 INFO:teuthology.repo_utils:Fetching wip-yuri4-testing-2019-03-20-2018-luminous from upstream
2019-03-21 15:24:20,012.012 INFO:teuthology.repo_utils:Resetting repo at /home/yuriw/src/github.com_ceph_ceph-c_wip-yuri4-testing-2019-03-20-2018-luminous to branch wip-yuri4-testing-2019-03-20-2018-luminous
2019-03-21 15:24:20,104.104 DEBUG:teuthology.suite.run:Suite rados in /home/yuriw/src/github.com_ceph_ceph-c_wip-yuri4-testing-2019-03-20-2018-luminous/qa/suites/rados
2019-03-21 15:24:20,104.104 INFO:teuthology.suite.build_matrix:Subset=35/999
Traceback (most recent call last):
File "/home/yuriw/teuthology/virtualenv/bin/teuthology-suite", line 11, in <module>
load_entry_point('teuthology', 'console_scripts', 'teuthology-suite')()
File "/home/yuriw/teuthology/scripts/suite.py", line 152, in main
return teuthology.suite.main(args)
File "/home/yuriw/teuthology/teuthology/suite/init.py", line 144, in main
run.prepare_and_schedule()
File "/home/yuriw/teuthology/teuthology/suite/run.py", line 343, in prepare_and_schedule
num_jobs = self.schedule_suite()
File "/home/yuriw/teuthology/teuthology/suite/run.py", line 494, in schedule_suite
build_matrix(suite_path, subset=self.args.subset, seed=self.args.seed)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 57, in build_matrix
mat, first, matlimit = _get_matrix(path, subset)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 67, in _get_matrix
mat = _build_matrix(path, mincyclicity=outof)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 146, in _build_matrix
fn)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 130, in _build_matrix
item=fn)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 146, in _build_matrix
fn)
File "/home/yuriw/teuthology/teuthology/suite/build_matrix.py", line 84, in _build_matrix
raise IOError('%s does not exist (abs %s)' % (path, os.path.abspath(path)))
IOError: /home/yuriw/src/github.com_ceph_ceph-c_wip-yuri4-testing-2019-03-20-2018-luminous/qa/suites/rados/singleton-bluestore/objectstore/bluestore.yaml does not exist (abs /home/yuriw/src/github.com_ceph_ceph-c_wip-yuri4-testing-2019-03-20-2018-luminous/qa/suites/rados/singleton-bluestore/objectstore/bluestore.yaml)

**Comment by ifed01** [comment]
@yuriw - thanks for testing. Fixed.

**Comment by yuriw** [comment]
[wip-yuri6-testing-2019-03-22-2012-luminous](https://trello.com/c/xiut3Zb9/565-wip-yuri6-testing-2019-03-22-2012-luminous)

**Comment by yuriw** [comment]
[wip-yuri3-testing-2019-03-28-1550-luminous](https://trello.com/c/Pmz5A7di/569-wip-yuri3-testing-2019-03-28-1550-luminous)

**Pull request merged**